pkg/k8s/cep.go

```diff
@@ -0,0 +1,363 @@
+		})
+}
+
+// RunK8sCiliumEndpointSync starts a controller that synchronizes the endpoint
+// to the corresponding k8s CiliumEndpoint CRD
+// CiliumEndpoint objects have the same name as the pod they represent
+func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoint) {
+	var (
+		endpointID     = e.ID
+		controllerName = fmt.Sprintf("sync-to-k8s-ciliumendpoint (%v)", endpointID)
+		scopedLog      = e.Logger(subsysK8s).WithField("controller", controllerName)
+		err            error
+	)
+
+	if option.Config.DisableCiliumEndpointCRD {
+		scopedLog.Warn("Not running controller. CEP CRD synchronization is disabled")
+		return
+	}
+
+	if !IsEnabled() {
+		scopedLog.Debug("Not starting controller because k8s is disabled")
+		return
+	}
+	k8sServerVer, err := GetServerVersion()
+	if err != nil {
+		scopedLog.WithError(err).Error("unable to retrieve kubernetes serverversion")
+		return
+	}
+	if !ciliumEPControllerLimit.Check(k8sServerVer) {
+		scopedLog.WithFields(logrus.Fields{
+			"expected": k8sServerVer,
+			"found":    ciliumEPControllerLimit,
+		}).Warn("cannot run with this k8s version")
+		return
+	}
+
+	ciliumClient, err := getCiliumClient()
+	if err != nil {
+		scopedLog.WithError(err).Error("Not starting controller because unable to get cilium k8s client")
+		return
+	}
+
+	// The health endpoint doesn't really exist in k8s and updates to it caused
+	// arbitrary errors. Disable the controller for these endpoints.
+	if isHealthEP := e.HasLabels(pkgLabels.LabelHealth); isHealthEP {
+		scopedLog.Debug("Not starting unnecessary CEP controller for cilium-health endpoint")
+		return
+	}
+
+	var (
+		lastMdl  *models.Endpoint
+		firstRun = true
+	)
+
+	// NOTE: The controller functions do NOT hold the endpoint locks
+	e.UpdateController(controllerName,
+		controller.ControllerParams{
+			RunInterval: 10 * time.Second,
+			DoFunc: func() (err error) {
+				// Update logger as scopeLog might not have the podName when it
+				// was created.
+				scopedLog = e.Logger(subsysK8s).WithField("controller", controllerName)
+
+				podName := e.GetK8sPodName()
+				if podName == "" {
+					scopedLog.Debug("Skipping CiliumEndpoint update because it has no k8s pod name")
+					return nil
+				}
+
+				namespace := e.GetK8sNamespace()
+				if namespace == "" {
+					scopedLog.Debug("Skipping CiliumEndpoint update because it has no k8s namespace")
+					return nil
+				}
+
+				mdl := e.GetModel()
+				if reflect.DeepEqual(mdl, lastMdl) {
+					scopedLog.Debug("Skipping CiliumEndpoint update because it has not changed")
+					return nil
+				}
+				k8sMdl := (*cilium_v2.CiliumEndpointDetail)(mdl)
+
+				cep, err := ciliumClient.CiliumEndpoints(namespace).Get(podName, meta_v1.GetOptions{})
+				switch {
+				// The CEP doesn't exist. We will fall through to the create code below
+				case err != nil && k8serrors.IsNotFound(err):
+					break
+
+					// Delete the CEP on the first ever run. We will fall through to the create code below
```

Comment:
The indentation for comments above these switch cases are a bit inconsistent (compare with line 289).